src/Playwright/Core/Page.cs

```diff
@@ -1042,33 +1062,28 @@ private void Channel_BindingCall(object sender, BindingCallEventArgs e)
             }
         }
 
-        private void OnRoute(Route route, IRequest request)
+        private async Task OnRouteAsync(Route route, IRequest request)
         {
-            foreach (var routeHandler in _routes.ToList())
+            var routeHandlers = _routes.Where(r => (r.Regex?.IsMatch(request.Url) == true) ||
+                    (r.Function?.Invoke(request.Url) == true)).ToList();
+            foreach (var routeHandler in routeHandlers)
             {
-                if ((routeHandler.Regex?.IsMatch(request.Url) == true) ||
-                    (routeHandler.Function?.Invoke(request.Url) == true))
+                if (routeHandler.WillExpire())
+                {
+                    _routes.Remove(routeHandler);
+                }
+                var handled = await routeHandler.HandleAsync(route).ConfigureAwait(false);
```

Comment:
Why was the `try` / `catch` removed from here? These tests would never complete:

```c#
[PlaywrightTest]
public async Task RouteException()
{
    await Page.RouteAsync("**/*", (route) =>
    {
        throw new System.Exception("test");
    });
    await Page.GotoAsync(Server.EmptyPage);
}

[PlaywrightTest]
public async Task RouteExceptionTask()
{
    await Page.RouteAsync("**/*", (route) =>
    {
       return Task.FromException(new System.Exception("test"));
    });
    await Page.GotoAsync(Server.EmptyPage);
}
```